File: n2f/geometry/build_box.py
"""Build box functions for sampling near fields."""

import numpy as np


def build_box(faces, x_min, x_max, y_min, y_max, z_min, z_max,
              x_pts, y_pts, z_pts, scale, plot_flag=False, for_plot=False):
    """
    Builds a box that encompasses the array over which will be computed the
    near fields on the selected faces and returns the near field sampling
    points position and the patches areas.

    Parameters
    ----------
    faces : ndarray of bool (6,)
        Boolean vector of the faces to select [XY face up (top), 
        XY face down (bottom), XZ face up, XZ face down,
        YZ face up, YZ face down]
    x_min, x_max, y_min, y_max, z_min, z_max : float
        Box dimensions
    x_pts, y_pts, z_pts : int
        Number of sampling points along x, y, z directions
    scale : float
        Reduction of the box dimensions in [m] that scales
        proportionally the box within the initial dimensions
    plot_flag : bool, optional
        Plots the sampling grid if True
    for_plot : bool, optional
        Allows proper box dimensions plot if True

    Returns
    -------
    box_pos : ndarray (3, N_pts)
        Positions of the sampling points or grid intersections
    box_n : ndarray (3, N_faces)
        Outwardly directed normal unit vectors to the faces
    dS : ndarray
        Surface patches areas
    m_size : ndarray
        Matrix sizes collected for each face selected
    """
    nbr_faces = np.sum(faces)
    
    x_vect = np.linspace(x_min, x_max, x_pts + 1)
    y_vect = np.linspace(y_min, y_max, y_pts + 1)
    z_vect = np.linspace(z_min, z_max, z_pts + 1)
    
    dx = abs(x_vect[0] - x_vect[1])
    dy = abs(y_vect[0] - y_vect[1])
    dz = abs(z_vect[0] - z_vect[1])
    
    if not for_plot:
        x_min = x_min * scale
        y_min = y_min * scale
        z_min = z_min * scale
        x_max = x_max * scale
        y_max = y_max * scale
        z_max = z_max * scale
        
        x_min_tmp = x_min + dx / 2
        y_min_tmp = y_min + dy / 2
        z_min_tmp = z_min + dz / 2
        x_max_tmp = x_max - dx / 2
        y_max_tmp = y_max - dy / 2
        z_max_tmp = z_max - dz / 2
        
        x_vect = np.linspace(x_min_tmp, x_max_tmp, x_pts)
        y_vect = np.linspace(y_min_tmp, y_max_tmp, y_pts)
        z_vect = np.linspace(z_min_tmp, z_max_tmp, z_pts)
    
    # Building box
    face_list = []
    
    if faces[0]:
        # XY face up
        pos_x, pos_y = np.meshgrid(x_vect, y_vect)
        pos_z = z_max * np.ones_like(pos_x)
        face_list.append({
            'pos_x': pos_x, 'pos_y': pos_y, 'pos_z': pos_z,
            'dS': dx * dy * np.ones_like(pos_x),
            'nVx': 0, 'nVy': 0, 'nVz': 1
        })
    
    if faces[1]:
        # XY face down
        pos_x, pos_y = np.meshgrid(x_vect, y_vect)
        pos_z = z_min * np.ones_like(pos_x)
        face_list.append({
            'pos_x': pos_x, 'pos_y': pos_y, 'pos_z': pos_z,
            'dS': dx * dy * np.ones_like(pos_x),
            'nVx': 0, 'nVy': 0, 'nVz': -1
        })
    
    if faces[2]:
        # XZ face up
        pos_x, pos_z = np.meshgrid(x_vect, z_vect)
        pos_y = y_max * np.ones_like(pos_x)
        face_list.append({
            'pos_x': pos_x, 'pos_y': pos_y, 'pos_z': pos_z,
            'dS': dx * dz * np.ones_like(pos_x),
            'nVx': 0, 'nVy': 1, 'nVz': 0
        })
    
    if faces[3]:
        # XZ face down
        pos_x, pos_z = np.meshgrid(x_vect, z_vect)
        pos_y = y_min * np.ones_like(pos_x)
        face_list.append({
            'pos_x': pos_x, 'pos_y': pos_y, 'pos_z': pos_z,
            'dS': dx * dz * np.ones_like(pos_x),
            'nVx': 0, 'nVy': -1, 'nVz': 0
        })
    
    if faces[4]:
        # YZ face up
        pos_y, pos_z = np.meshgrid(y_vect, z_vect)
        pos_x = x_max * np.ones_like(pos_y)
        face_list.append({
            'pos_x': pos_x, 'pos_y': pos_y, 'pos_z': pos_z,
            'dS': dy * dz * np.ones_like(pos_y),
            'nVx': 1, 'nVy': 0, 'nVz': 0
        })
    
    if faces[5]:
        # YZ face down
        pos_y, pos_z = np.meshgrid(y_vect, z_vect)
        pos_x = x_min * np.ones_like(pos_y)
        face_list.append({
            'pos_x': pos_x, 'pos_y': pos_y, 'pos_z': pos_z,
            'dS': dy * dz * np.ones_like(pos_y),
            'nVx': -1, 'nVy': 0, 'nVz': 0
        })
    
    # Stack all positions and normals
    pos_x_all = np.concatenate([f['pos_x'].flatten() for f in face_list])
    pos_y_all = np.concatenate([f['pos_y'].flatten() for f in face_list])
    pos_z_all = np.concatenate([f['pos_z'].flatten() for f in face_list])
    box_pos = np.vstack([pos_x_all, pos_y_all, pos_z_all])
    
    n_vx_all = np.concatenate([np.full_like(f['pos_x'].flatten(), f['nVx']) for f in face_list])
    n_vy_all = np.concatenate([np.full_like(f['pos_y'].flatten(), f['nVy']) for f in face_list])
    n_vz_all = np.concatenate([np.full_like(f['pos_z'].flatten(), f['nVz']) for f in face_list])
    box_n = np.vstack([n_vx_all, n_vy_all, n_vz_all])
    
    dS = np.concatenate([f['dS'].flatten() for f in face_list])
    
    m_size = np.array([f['pos_x'].shape for f in face_list])
    
    return box_pos, box_n, dS, m_size

File: matlabLib/buildBox.m
% Constructs a rectangular bounding box surface for near-field sampling
%
% Creates rectangular faces with optional selective inclusion for N2F transformations.
% Enables flexible bounding surface choices (full box, partial enclosure, single plane)
%
% FACE INDEXING: faces = [faceTop, faceBottom, faceUp, faceDown, faceLeft, faceRight]
%   All indices use 1 = include, 0 = exclude
%   faceTop = XY plane at z=zMax (outward normal +z)
%   faceBottom = XY plane at z=zMin (outward normal -z)
%   faceUp = XZ plane at y=yMax (+y normal)
%   faceDown = XZ plane at y=yMin (-y normal)
%   faceLeft = YZ plane at x=xMin (-x normal)
%   faceRight = YZ plane at x=xMax (+x normal)
%
%   Example: [1 1 1 1 1 1] = all six faces (closed box)
%   Example: [0 0 1 1 1 1] = four sides only (open top/bottom)
%   Example: [1 0 0 0 0 0] = single face (for MoM problems)
%
% [boxPos, boxN, dS, mSize] = buildBox(faces, xMin, xMax, yMin, yMax, zMin, zMax, ...
%   xPts, yPts, zPts, scale, plotFlag, forPlot)
%
% IN: faces = [6x1] binary flag vector selecting faces as described above
%     xMin, xMax, yMin, yMax, zMin, zMax = box boundaries [m]
%     xPts, yPts, zPts = number of sampling points along each edge direction
%     scale = box contraction factor [m]; final dimensions = original - scale
%             Used for positioning inside room or near geometry
%     plotFlag = 1 to visualize the sampling grid
%     forPlot = 1 to ensure patches align for proper 3D visualization
%
% OUT: boxPos = Cartesian coordinates of sampling points (3 x nTotalPts)
%      boxN = outward normal unit vectors to each point (3 x nTotalPts)
%      dS = area per patch element [m^2] (1 x nTotalPts)
%           Uniform spacing: dS(i) = (dX * dY) or (dX * dZ) or (dY * dZ)
%      mSize = mesh metadata: dimensions, statistics, face tallies
%
% GEOMETRY: Each face is a regular rectangular mesh
%   XY face: dX = (xMax-xMin)/(xPts-1), dY = (yMax-yMin)/(yPts-1)
%   Outward normals always point AWAY from box interior
%
% Laurent Ntibarikure
%               the sampling patches if forPlot is asserted
%      boxN = outwardly directed normal unit vectors to the faces
%      dS = surface patches areas
%      mSize = matrices sizes collected in a 3D matrix of depth the number
%              of faces selected
%
% Laurent Ntibarikure 
function [boxPos, boxN, dS, mSize] =...
    buildBox(faces, xMin, xMax, yMin, yMax, zMin, zMax,...
    xPts, yPts, zPts, scale, plotFlag, forPlot)

nbrFaces = length(find(faces));
 
xVect = linspace(xMin, xMax, xPts+1);
yVect = linspace(yMin, yMax, yPts+1);
zVect = linspace(zMin, zMax, zPts+1);

dx = abs(xVect(1)-xVect(2));
dy = abs(yVect(1)-yVect(2));
dz = abs(zVect(1)-zVect(2));

if nargin>12 && forPlot
else
  xMin = xMin * scale;
  yMin = yMin * scale;
  zMin = zMin * scale;
  xMax = xMax * scale;
  yMax = yMax * scale;
  zMax = zMax * scale;
  
  xMinTmp = xMin + dx/2;
  yMinTmp = yMin + dy/2;
  zMinTmp = zMin + dz/2;
  xMaxTmp = xMax - dx/2;
  yMaxTmp = yMax - dy/2;
  zMaxTmp = zMax - dz/2;

  xVect = linspace(xMinTmp, xMaxTmp, xPts);
  yVect = linspace(yMinTmp, yMaxTmp, yPts);
  zVect = linspace(zMinTmp, zMaxTmp, zPts);
end
%% Building box
k=0;
if faces(1)
  % XY face up
  k=k+1;
  [face(k).pos_x, face(k).pos_y] = meshgrid(xVect, yVect);
  face(k).pos_z = (zMax)*ones(size(face(k).pos_x));
  face(k).dS = dx*dy*ones(size(face(k).pos_x));
  face(k).nVx = 0;
  face(k).nVy = 0;
  face(k).nVz = 1;
end

if faces(2)
  % XY face down
  k=k+1;
  [face(k).pos_x, face(k).pos_y] = meshgrid(xVect, yVect);
  face(k).pos_z = (zMin)*ones(size(face(k).pos_x));
  face(k).dS = dx*dy*ones(size(face(k).pos_x));
  face(k).nVx = 0;
  face(k).nVy = 0;
  face(k).nVz = -1;
end

if faces(3)
  % XZ face up
  k=k+1;
  [face(k).pos_x, face(k).pos_z] = meshgrid(xVect, zVect);
  face(k).pos_y = (yMax)*ones(size(face(k).pos_x));
  face(k).dS = dx*dz*ones(size(face(k).pos_x));
  face(k).nVx = 0;
  face(k).nVy = 1;
  face(k).nVz = 0;
end

if faces(4)
  % XZ face down
  k=k+1;
  [face(k).pos_x, face(k).pos_z] = meshgrid(xVect, zVect);
  face(k).pos_y = (yMin)*ones(size(face(k).pos_x));
  face(k).dS = dx*dz*ones(size(face(k).pos_x));
  face(k).nVx = 0;
  face(k).nVy = -1;
  face(k).nVz = 0;
end

if faces(5)
  % YZ face up
  k=k+1;
  [face(k).pos_y, face(k).pos_z] = meshgrid(yVect, zVect);
  face(k).pos_x = (xMax)*ones(size(face(k).pos_y));
  face(k).dS = dy*dz*ones(size(face(k).pos_y));
  face(k).nVx = 1;
  face(k).nVy = 0;
  face(k).nVz = 0;
end

if faces(6)
  % YZ face down
  k=k+1;
  [face(k).pos_y, face(k).pos_z] = meshgrid(yVect, zVect);
  face(k).pos_x = (xMin)*ones(size(face(k).pos_y));
  face(k).dS = dy*dz*ones(size(face(k).pos_y));
  face(k).nVx = -1;
  face(k).nVy = 0;
  face(k).nVz = 0;
end
%% Total number of samples
totalNbrSamples = 0;
if plotFlag
  figure;
  colormap(getColorMap);
end
for k=1:nbrFaces
  totalNbrSamples = totalNbrSamples + ...
    size(face(k).pos_x,1) * ...
    size(face(k).pos_x,2);
  if plotFlag
    mesh(face(k).pos_x,face(k).pos_y,face(k).pos_z, ...
      'FaceAlpha',0,'EdgeAlpha',1, 'EdgeColor', getColorMap);
    hold on; axis('equal');
  end
end
%%
boxPos = zeros(3,totalNbrSamples);
dS = zeros(1,totalNbrSamples);
boxN = zeros(3,totalNbrSamples);
mSize = zeros(1,2,nbrFaces);
valTot = 0;
for k=1:nbrFaces
  val = size(face(k).pos_x(:),1);
  mSize(:,:,k) = size(face(k).pos_x);
  boxPos(1,valTot+(1:val))= (face(k).pos_x(:)).';
  boxPos(2,valTot+(1:val))= (face(k).pos_y(:)).';
  boxPos(3,valTot+(1:val))= (face(k).pos_z(:)).';
  dS(1,valTot+(1:val))= (face(k).dS(:)).';
  boxN(1,valTot+(1:val))= face(k).nVx(:)*ones(1,val);
  boxN(2,valTot+(1:val))= face(k).nVy(:)*ones(1,val);
  boxN(3,valTot+(1:val))= face(k).nVz(:)*ones(1,val);
  valTot = valTot + val;
end